src/book/03-guides/01-main-codebase/02-nuclear-tutorial.mdx

```diff
@@ -0,0 +1,312 @@
+
+In `TestModule.yaml` (in your text editor, not in the terminal since this is not persistent), add in the following lines
+
+```yaml
+# Each time the Ping Pong messages are sent, a count is incremented and the increment is of the following size
+increment: 2
+```
+
+Next we need to read it in the module. Go to `TestModule.hpp`. We will make a variable to store this in. Inside the config struct on line 11, add in
+
+```cpp
+int increment = 0;
+```
+
+Always initialise your variables. Save and go to `TestModule.cpp`. Add in the following lines to read the configuration value and print a message to let the user know the increment size.
+
+```cpp
+config.increment = cfg["increment"].as<int>();
+log<NUClear::INFO>("Increment is of size", config.increment);
+```
+
+Save, recompile and check that the program prints the message.
+
+```bash
+./b build
+./b run testprogram
+```
+
+Make sure you see your INFO log and then lets move onto making the messages that we need.
+
+### Protobuf Messages
+
+The idea of the program is that one reactor will send a `Ping` message when it receives a `Pong` message, and the other reactor will send a `Pong` message when it receives a `Ping` message.
+
+We will need to create these messages. In `shared/message` create a new file `PingPong.proto`. Put the follow code into the file.
+
+```proto
+syntax = "proto3";
+
+package message;
+
+message Ping {
+    uint32 count = 1;
+}
+
+message Pong {
+    uint32 count = 1;
+}
+```
+
+There is an integer in each of these messages. This is the count that will be incremented each time a new `Ping` or `Pong` message is sent. The number it is assigned, `1`, is not the value but its position in the Protobuf message. Save this file and head back over to `TestModule.cpp`.
+
+Up at the top of the file, include the protobuf message we just created. Note that these messages are compiled into C++ code, so include the compiled header.
+
+```cpp
+#include "message/PingPong.hpp"
+```
+
+Make sure it compiles and still runs successfully. Because we added files, we will need to configure again.
+
+```bash
+./b configure
+./b build
+./b run testprogram
+```
+
+If this works, you can move onto the next section where we make our reactors.
+
+### Reactors
+
+Reactors are chunks of code that will run given certain conditions are true. We will make two reactors, one that runs reactions whenever it gets a `Pong` message, and the other when it gets a `Ping` message.
+
+Head back to `TestModule.cpp`. Underneath the Configuration reactor, add in the following reactor
+
+```cpp
+on<Trigger<Ping>>().then([this](const Ping& ping) {
+    // Some code!
+});
+```
+
+`Trigger` means that it will run when it a `Ping` message is available. We get access to this message via the `ping` parameter. Let's add in the other reactor.
```

Comment:
```suggestion
`Trigger` means that it will run when it a `Ping` message is available. We get access to this message via the `ping` parameter. Let's add in the other reaction.
```